### DISCUSSION TITLE
Snapserver/client issues

### DISCUSSION THREAD

**Discussion by mailman-fin** [rern/rAudio#1045] [General]
Hello,

Totally newbie here with rAudio-1 (also not so familiar with linux), but this should be what I have been looking for. And that is Multiroom streaming from Spotify.

So I have 2 Rasp's (called rAudio1 and rAudio2) where I have installed rAudio-1 successfully.
Spotify renderer also works as it should. I can stream Spotify from cell phone/pc/whatever to both Rasp's separately. And as said, Multiroom is what I want.

rAudio1 (192.168.1.19); I did enable SnapServer (also SnapClient and Spotify renderer enabled on this Rasp).
I can still stream Spotify just ok for this one. I can also "connect" this rAudio1 Snapclient to itself pressing "connect" button in main screen after "Networks" button.

rAudio2 (192.168.1.18); I did enable SnapClient and Spotify renderer. I can stream to this Rasp just fine.
If I try to press "connect" button in main screen after "Networks" button, I just get answer "Snapcast server not available".

I can ping each Rasp's in both ways, with host name and IP.

rAudio1 SnapServer log looks like this;
snapcast 0.26.0-1
snapserver.service - Snapcast server
     Loaded: loaded (/usr/lib/systemd/system/snapserver.service; enabled; preset: disabled)
     Active: active (running) since Tue 2022-09-27 21:23:30 EEST; 1h 17min ago
       Docs: man:snapserver(1)
   Main PID: 347 (snapserver)
      Tasks: 5 (limit: 1062)
        CPU: 1min 18.419s
     CGroup: /system.slice/snapserver.service
             └─347 /usr/bin/snapserver --logging.sink=system --server.datadir=/var/lib/snapserver
Sep 27 22:21:12 rAudio1 snapserver[347]: Removing 1 inactive session(s), active sessions: 2
Sep 27 22:21:12 rAudio1 snapserver[347]: ControlServer::NewConnection: 192.168.1.25
Sep 27 22:21:23 rAudio1 snapserver[347]: StreamServer::NewConnection: 192.168.1.18
Sep 27 22:21:23 rAudio1 snapserver[347]: Hello from b8:27:eb:f1:ef:7b, host: rAudio2, v0.26.0, ClientName: Snapclient, OS: Arch Linux ARM, Arch: armv6l, Protocol version: 2
Sep 27 22:21:23 rAudio1 snapserver[347]: Error reading message header of length 0: End of file
Sep 27 22:21:23 rAudio1 snapserver[347]: onDisconnect: b8:27:eb:f1:ef:7b
Sep 27 22:21:23 rAudio1 snapserver[347]: StreamSession write error (msg length: 0): Broken pipe
Sep 27 22:21:23 rAudio1 snapserver[347]: Error in socket shutdown: Transport endpoint is not connected
Sep 27 22:23:11 rAudio1 snapserver[347]: State changed: default, state: playing => idle
Sep 27 22:23:11 rAudio1 snapserver[347]: onStateChanged (default): idle

192.168.1.25 is my workstation.
And for some reason I cannot get anymore anything from 192.168.1.18 to the log.

In rAudio2 (.18), I have tried to disable/enable SnapClient many times, but nothing...

Also I have enabled SnapServer in http://192.168.1.19:1780/ and there I can of course only see rAudio1, but the controls wont work when streaming Spotify.
When tested with preconfigured Webradios, noticed that those did run in my PC and rAudio1 just ok. AND Webcontrols did work too. So somehow Snapserver seems to work, I assume?

Really dont know how to go forward.
Yesterday evening googling did not help at all, and today I did reinstall both Rasp's from scratch, but still same issues.

Any help would be nice.

Thanks.

**Comment by skycamel** [comment]
Just a wild guess, but I think that it doesn't make any sense to activate and snapcast server and snapcast client on the same rAudio.
Try to only activate snapcast server on rAudio1 (and spotify) which will be the master. On the rAudio2 you only activate snapcast client will then be, well, your client. I don't think you even need to activate spotify on rAudio2 because you will be streaming your spotify over snapcast from rAudio1 to rAudio2. Again, just a guess ;-)
Cheers

**Reply by mailman-fin** [reply]
Thanks for the suggestions, but actually tried that one too, without any luck. Just forget to tell it here :/
Guide says that Snapcast server and client can work in same device.
Also tried without Spotify too.

**Comment by mailman-fin** [comment]
Anyone? :)

Seems that this community is quite quiet...

**Reply by mailman-fin** [reply]
Noticed quite new Release of rAudio-1 so tried again today, from the scratch.
But no, still I cannot get Snapcasting to work. Separately both Rasp's run just ok

Dummy instructions? I just cannot understand what I am really doing wrong...

**Reply by skycamel** [reply]
Well if it's any consolation, me too I'm waiting for a solution. I have my "main" rAudio with snapcastserver running but my "client" rAudio doesn't seem to be able to connect the snapcastclient. Other snapcast client seem to be able to connect. Hoping too one day there will be a solution

**Reply by f0rdprefect** [reply]
Hi there!

I would like to get started with snapcast as well. Since raspberry pi's are kind of expensive right now I would use some old cellphones to test the streaming.

However I did have quite some mixed results in even getting the server up and running. Am I missing specific documentation? I tried just to flip the switch and connect to the snapcast server on rAudio (I just have one) without any luck. 

Then I tried `systemctl start snapserver` which starts the server and I can connect with a snapcast client. But no audio so far.

Any pointers welcome at this point ;)

**Reply by f0rdprefect** [reply]
Meanwhile....

If nothing is playing, snapclient enabled in the GUI I can start snapclient from the commandline and it will pick up the stream from the host specified from the command line. I tried to set this as an option in `/etc/default/snapclient` but that gets overwritten by some logic when disabling/enabling snapclient from the GUI....

**Comment by f0rdprefect** [comment]
I guess I have a working solution to the original question figured out. 

I found that the systemd unit for snapclient has a `-` before the path to Environmentfile. 

I have not found any documentation what that minus means. But before removing it, I think the parameters in that file are simply not applied.

Consequently I added in `/etc/defaults/snapclient.conf`   ` --host <mysnapserverip>` to the options. This already allowed me to start and stop snapclient from the command line and connect to the specified server. If nothing was blocking the soundcard on the raudio one it should play the snapcast stream.

Furthermore I hacked `/srv/http/bash/snapcast.sh`. Instead of tusting that a second snapclient would connect to determine the IP address I use this in line 18:

`serverip=$( timeout 0.2 journalctl -f -u snapclient | awk '/Connected to/ {print $NF}' )`

with that I can start the snapclient also from the GUI. Stop seems to be a bit trickier.

To test the spotify stream I found this docker container / command:

`docker run -d --name snapserver --net host -e DEVICE_NAME=Snapcast mazzolino/librespot-snapserver`

Maybe at least the change in snapcast.sh and the change in the systemd unit can make it into the official build? That would be great!

Btw if one opens the settings for the snapclient the parameters in `/etc/default/snapclient` get overwritten!
